Case import history

dannyroberts opened the pull request:
Currently on staging:

<img width="1172" alt="screen shot 2016-12-07 at 2 32 27 pm" src="https://cloud.githubusercontent.com/assets/137212/20983333/0e7b4a84-bc8a-11e6-81e3-ea31bf07914d.png">

This displays below **Select a file to upload** section on case importer landing page currently, but could be convinced to place somewhere else. Currently supported features:

- Shows 10 most recent uploads (including uploads in progress)
- Shows progress of any active upload in the project space (including the one you just uploaded, or multiple uploads running at the same time)
- As long as any of the uploads shown are in progress, it will show you the progress live. (It currently stops polling once all displayed uploads are complete, so you'd have to refresh the page to see new ones.)
- Shows full success message and warnings that you'd see on the polling page
- Shows case type, name of the user who uploaded it, and time (in your chosen timezone)

Working with @sheelio on where to go with this next, but wanted to open this up for review and comments.

dannyroberts committed:
change assert_doc_lists_equal to assert_doc_sets_equal

dannyroberts committed:
add CaseUploadRecord and url to fetch its json

dannyroberts committed:
make soil use cononical uuid strings (with hyphens)

dannyroberts committed:
show case upload history on case importer landing page

dannyroberts committed:
add user to upload history

dannyroberts committed:
add case_type to upload history

dannyroberts committed:
add succcess message to upload history

dannyroberts committed:
fix template for no recent uploads

dannyroberts commented:
@czue buddy

benrudolph commented:
oh wow this is sweet 😮

czue commented:
yeah this is awesome!

snopoke commented on corehq/apps/users/dbaccessors/couch_users.py:
```diff
@@ -14,3 +14,11 @@ def get_user_id_by_username(username):
     return None
+
+
+def get_display_name_for_user_id(domain, user_id, default=None):
```
caching?

dannyroberts replied:
Good eye. It's not obvious from this code, but `CouchUser.get_by_user_id` is already cached, I believe with enough sophistication that it's properly cleared when a user is saved, so caching this function is at best redundant and at worst would lead to the minor stale cache issues.

snopoke commented on corehq/apps/case_importer/tracking/models.py:
```diff
@@ -0,0 +1,41 @@
+        return get_task(self.task_id)
+
+
+class CaseUploadJSON(jsonobject.JsonObject):
```
Interested to hear why you chose this approach to producing JSON instead of having a `to_json` method on the `CaseUploadRecord` that either just creates a dict or uses a serializer to create the dict.

dannyroberts replied:
I had two objectives in writing this, though I'm not too tied to it and it's definitely more verbose than the alternatives.

1. have an explicitly-defined read-only model for `get_case_uploads` to return, rather than a queryset which lets the caller do all sorts of things to the underlying SQL tables.
2. have a well-defined schema for the JSON object to be passed on for use in javascript

(2) ended up not working out since I later realized I need to do some further transformations before passing it on for use in javascript, but 1 is still a pro. It's definitely adds some redundancy, so I may scrap it anyway for the simpler option, or keep some aspect of it but simplify it somehow.

snopoke commented:
Looks great! I think the main goal for this is to be able to run the same import multiple times without the need to re-configure it. In that case we'd need to store the import config somewhere. I guess we might also want to be able to permanently save a config for reuse (instead of just the last 10). 

The workflow that comes to mind is that after running an import you get the option to make it an import template (button on the history). Then when doing a new import you could select a template instead of going through all the config.

@sheelio can correct me if I'm off base.

dannyroberts commented:
@snopoke thanks for the review. As to your last comment, that's one direction this feature could go in, but there are a number of other offshoots we're considering that are all enabled by having some sort of record of previous uploads. That includes

- getting access to the original file used for an upload which can help a lot with debugging, by removing uncertainty as to what was done in a particular upload
- maintaining links to the cases touched by an upload and/or the forms used to touch them
- which further opens the door for a feature to revert an upload (by reverting all the forms involved)

as well as just the ability to see what uploads are already in progress so you don't get confused and double-upload something, or two users don't step on each others' toes by doing similar uploads at the same time, etc.

Each of those things requires adding more columns to what's stored for each upload, as does storing the config, but they all start with this basic feature of having a record of past uploads, so I'm just starting with this and then working with @sheelio to prioritize the other add-ons.

The config thing is something we've considered, but it also does have some open questions, like how to make sure the new excel file you choose is compatible with the config, etc. that aren't problems when you're building the config off the excel file like is currently done. That made it enough of an uncertain feature that I'm pushing it down the road a bit and will probably work on some lower-hanging fruit first.

dannyroberts commented:
@snopoke I forgot to say re: last 10, though you may have picked this up. The case upload records are persisted (in postgresql) forever at this point, and I just made the front end arbitrarily ask for the last 10 and punted any design decisions related to the UI for getting more, etc. So yeah, for now just showing the last 10, but assuming that longer term we'll want to change that and that it won't be too hard to change.

dannyroberts committed:
change wording in case import history

dannyroberts commented:
@snopoke / any reviewer: I have a number of branches building on this one, and I'm interested in hearing your opinions about rollout/review plan. Some options

- QA and merge this branch, releasing this first-pass to our users, and then PR the other branches to be reviewed individually and merged and deployed as they can be
- combine those other branches into this one and deploy them as a single change-set, rather than giving users a few new features every day

I _think_ the learning curve for users for this change isn't bad at all, so I'd weight being deliberate about the changes we expose lower than I would for app builder, but still could be a consideration. I'm fine with either way, though the long-running branch may get more and more annoying for reviewers to keep track of.

dannyroberts commented:
OK, I tested case importer on staging, as well as fixture upload and user upload (to validate my touching soil internals didn't break anything). Open question above about rollout, but from a QA perspective this is g2g

snopoke commented on corehq/apps/case_importer/templates/case_importer/import_cases.html:
```diff
@@ -2,6 +2,53 @@
 {% load i18n %}
 
+{% block js-inline %}
+<script id="caseUploadsResource" type="text/plain">{% url 'case_importer_uploads' domain %}</script>
```
This is interesting, don't think I've ever seen this pattern before. Why is it better than just assigning the URL to a var directly?

dannyroberts replied:
Yeah good question. If I put `"{% url ''case_importer_uploads' domain %}"` inline, then it's escaping for an html context but putting in a string-in-javascript context, so there are a number of things that can go wrong, even though it's rare. Now that I think of it though, I believe `{% url ''case_importer_uploads' domain as url %}{{ url|JSON }}` is the simplest way to do this correctly, and I think that's a better convention, so I'll switch to that. Thanks for asking

snopoke commented on corehq/apps/case_importer/templates/case_importer/import_cases.html:
```diff
@@ -48,4 +95,71 @@
+                    <!--/ko-->
+                </td>
+                <td class="col-md-1">
+                    <!--ko text: case_type--><!--/ko-->
```
why use a comment binding there instead of a text binding on the `td` or on a `span`?

dannyroberts replied:
Yeah no particular reason. Sounds like you think we should prefer data-bind on an element over comment bindings, it's certainly more common in our code. I can change this.

snopoke commented on corehq/apps/case_importer/templates/case_importer/partials/ko_import_status.html:
```diff
@@ -0,0 +1,64 @@
+
+        <!--ko if: result.created_count > 0-->
+            <p>
+                {% blocktrans with created_count='result.created_count' %}
```
If these are available in the template (`result.created_count`) why use knockout to determine whether or not to show them? Especially since they don't seem to get updated by any JS if the value changes.

dannyroberts replied:
Sorry this is confusing. This line sets the django template variable `created_count` to the string `'result.created_count'`, which is then used below. The point was just remove as much logic as possible from the thing inside blocktrans, but it's still pretty awkward. Not sure if there's a much better way to do this.

snopoke replied:
And what's bad about just having this inside the blocktrans:
```
<strong data-bind="text: result.created_count"></strong>
```

snopoke replied:
Yea, I see why you do that, seems OK I guess

snopoke commented on corehq/apps/case_importer/templates/case_importer/partials/ko_import_status.html:
```diff
@@ -0,0 +1,64 @@
+        <!--ko if: result.created_count > 0-->
+            <p>
+                {% blocktrans with created_count='result.created_count' %}
+                    <strong data-bind="text: {{ created_count }}"></strong> rows did
```
Also confused by this, why use Knockout data-bind with a static value instead of just putting the value in?

dannyroberts replied:
Yeah the django template evaluates this to

```html
<strong data-bind="text: result.created_count"></strong>
```

so it's still a dynamic binding. The idea is that if I change the structure of `result` somehow and want to edit where we're getting that value from, I can change it outside of the blocktrans block, so that the string doesn't have to be translated again. Agree it's pretty awkward and open to better suggestions.

snopoke commented on corehq/apps/case_importer/templates/case_importer/partials/ko_import_status.html:
```diff
@@ -0,0 +1,64 @@
+    <div class="alert alert-warning">
+        <ul class="list-unstyled">
+        <!--ko foreach: _.pairs(result.errors).map(_.partial(_.object, ['error','columns']))-->
+            <!--ko foreach: _.pairs(columns).map(_.partial(_.object, ['column', 'value']))-->
```
can this not be pushed into the JS rather than having in the ko comment? It would make it easier to test at a later stage if necessary.

dannyroberts replied:
yeah that's a good point. I can do this.

snopoke commented on corehq/apps/case_importer/tracking/models.py:
```diff
@@ -0,0 +1,41 @@
+
+
+class CaseUploadRecord(models.Model):
+    domain = models.CharField(max_length=256)
```
should add an index to this.

dannyroberts replied:
yeah maybe even on (domain, created)

snopoke commented on corehq/apps/case_importer/tracking/views.py:
```diff
@@ -0,0 +1,39 @@
+        tz = get_timezone_for_request()
+        task_status = get_task_status(get_task(case_upload.task_id))
+        return {
+            'created': ServerTime(case_upload.created).user_time(tz).ui_string(),
```
nice

snopoke reviewed (commented):
Nothing blocking and fine with pushing this out now. Will let you merge in case any of my questions prompt changes.

czue commented:
looks like @snopoke is doing a better review than I would so will let him review/merge. let me know if anything needed though

dannyroberts commented on corehq/apps/case_importer/templates/case_importer/import_cases.html:
```diff
@@ -2,6 +2,53 @@
+<script id="caseUploadsResource" type="text/plain">{% url 'case_importer_uploads' domain %}</script>
+<script>
+(function () {
+    var caseUploadResource = $('#caseUploadsResource').text();
```
Probably definition should be in a separate file and binding should be on the page. That's the closest thing we've come to to a convention for this, so you're right I ought to adhere to it.

dannyroberts commented:
@snopoke thanks for all the great comments. I think what I'll do is merge this now, PR the other changes I've done on top of this, and then incorporate all your feedback on a branch on top of that, just because that'll be a lot simpler for me than making the changes here and then merging it with the changes I've made on top of this.

dannyroberts commented:
@snopoke addressed your PR feedback in these commits: https://github.com/dimagi/commcare-hq/compare/case-import-history-phase2...case-import-history-phase-1-feedback, which are on top of this PR: https://github.com/dimagi/commcare-hq/pull/14178. You should be able to comment on the commits even though it's not in PR yet.

I added a little tooling around urls that's pretty generic and useful in other places; I was really not liking the redundancy of passing around urls, but am open to it being not the right fix.

snopoke commented:
@dannyroberts I think that's a pretty good approach but since you're proposing a new way of doing url's in JS I think it's worth sending to a wider audience (architecture review / JS team)

dannyroberts commented:
@snopoke cool, I just took it to slack#future-of-js-in-hq and it got a positive reception; I'll also ping them on the PR when I open it
